src/enkf/controlvec.f90

```diff
@@ -309,11 +309,21 @@ subroutine write_control(no_inflate_flag)
          print *,'--------------'
       endif
       ! gather ensmean increment on root.
-      do ne=1,nanals_per_iotask
-         call mpi_reduce(grdin(:,:,nb,ne), grdin_mean_tmp, npts*ncdim, mpi_real4,&
-                         mpi_sum,0,mpi_comm_io,ierr)
-         if (nproc == 0) grdin_mean(:,:,nb,1) = grdin_mean(:,:,nb,1) + grdin_mean_tmp
-      enddo
+      if (real(npts)*real(ncdim) < 2**32/2. - 1) then
```

Comment:
Need to specify numeric precision for `2**32/2. - 1` (GSI code  standard)

Reply:
I will fix this.

Reply:
@TingLei-NOAA 's comment prompted me to add print statements to see which branch of the `if-then` block is entered.  Much to my surprise, the `else` branch was always entered despite the fact that the product of `npts` and `ncdim` is `7,526,424`.     This shouldn't be the case since `2*32/2.-1` is `2,147,483,647`

I wrote a simple test code
```
program test
  integer(4),parameter:: one_i4 = 1_4
  integer(4),parameter:: two_i4 = 2_4
  real(8),parameter:: one_r8= 1.0_8
  real(8),parameter:: two_r8 = 2.0_8
  real(8):: term

  term = 2**32
  write(6,*)'2**32 is is ',term

  term = 2**32/2.
  write(6,*)'2**32/2. is ',term

  term=2**32/2. - 1
  write(6,*)'2**32/2. - 1 is ',term

  term=two_i4**32/two_r8 - one_i4
  write(6,*)'two_i4**32/two_r8 - one_i4 is ',term

  term=two_r8**32/two_r8 - one_r8
  write(6,*)'two_r8**32/two_r8 - one_r8 is ',term

  term=two_r8**32/two_i4 - one_i4
  write(6,*)'two_r8**32/two_i4 - one_i4 is ',term          

end program test
```
Execution of the resulting `./a.out` yields
```
 2**32 is is   0.000000000000000E+000
 2**32/2. is   0.000000000000000E+000
 2**32/2. - 1 is   -1.00000000000000     
 two_i4**32/two_r8 - one_i4 is   -1.00000000000000     
 two_r8**32/two_r8 - one_r8 is   2147483647.00000     
 two_r8**32/two_i4 - one_i4 is    2147483647.00000  
```
This makes sense.  `2**32` is `4,294,967,296`.   But `2**32` in the code is integer arithmetic.  The maximum value for an integer(4) is `2,147,483,647`.

If we want the logical tests in `loadbal.f90` and `controlvec.f90` to work, we need to use real arithmetic to compute `2*32/2. - 1`.    It seems that as is the code is not working as intended.

What do you think, @jswhit2 ?

Reply:
I've now updated the if/else checks to use ```2_r_kind**32/2_r_kind - 1_r_kind```

Reply:
Recompile the code with the changes committed at dfd4e7dcc.   Unfortunately, a rerun `enkf.x` for the test case yields the following printout in `stdout`
```
ens. mean anal. increment min/max  v   -271.803466797        325.930572510
ens. mean anal. increment min/max  tv   -466.174316406        428.788879395
ens. mean anal. increment min/max  q   -198.692459106        204.148773193
ens. mean anal. increment min/max  oz   -915.857604980        1035.82275391
ens. mean anal. increment min/max  ps   -20.3442897797        19.5052604675
```
The control built from the head of develop has
```
ens. mean anal. increment min/max  u   -8.92652511597        11.2560396194
ens. mean anal. increment min/max  v   -8.99013900757        7.72295475006
ens. mean anal. increment min/max  tv   -4.58017873764        4.80190753937
ens. mean anal. increment min/max  q  -0.546840786934       0.393992841244
ens. mean anal. increment min/max  oz  -0.987113935480E-06   0.857482405081E-06
ens. mean anal. increment min/max  ps   -2.43661499023        2.57639765739
```
A query of the netcdf analysis increment files from the test code confirms that the min/max increments are unrealistically large.

Something isn't correct when execution enters the `if` branch of the 2*32 `if-then` block.  I am running on WCOSS2.   Does the code execute as expected on Hera or Orion?

Reply:
Thank you for identifying this problem @RussTreadon-NOAA.  I will investigate further as soon as I get a free moment.  When the regression tests passed, the code was always going into the other branch of the if/else (the one intended for very large grids), is that correct?

Reply:
Yes @jswhit2 , I added print statements in each branch of the if-then block.   

- When we had `2**32/2. - 1`, execution always entered the else branch.    Analysis increments look reasonable.   
- After moving to `r_kind` arithmetic execution now enters the if branch.    Analysis increments are unphysically large.